xarray/core/datatree.py

```diff
@@ -393,31 +429,58 @@ def __init__(
         --------
         DataTree.from_dict
         """
-
-        # validate input
         if children is None:
             children = {}
-        ds = _coerce_to_dataset(data)
-        _check_for_name_collisions(children, ds.variables)
 
         super().__init__(name=name)
 
-        # set data attributes
-        self._replace(
-            inplace=True,
-            variables=ds._variables,
-            coord_names=ds._coord_names,
-            dims=ds._dims,
-            indexes=ds._indexes,
-            attrs=ds._attrs,
-            encoding=ds._encoding,
-        )
-        self._close = ds._close
+        # set tree attributes
+        self._children = {}
+        self._parent = None
+        self._set_node_data(data)
 
-        # set tree attributes (must happen after variables set to avoid initialization errors)
-        self.children = children
+        # finalize tree attributes
+        self.children = children  # must set first
         self.parent = parent
 
+    def _set_node_data(self, data: Dataset | DataArray | None) -> None:
+        ds = _coerce_to_dataset(data)
+
+        # set node data attributes
+        self._local_variables = ds._variables
+        self._local_coord_names = ds._coord_names
+        self._local_dims = ds._dims
+        self._local_indexes = ds._indexes
+        self._attrs = ds._attrs
+        self._encoding = ds._encoding
+        self._close = ds._close
+
+        # setup inherited node attributes (finalized by _post_attach)
+        self._inherited_coord_names = set()
+
+    def _pre_attach(self: DataTree, parent: DataTree, name: str) -> None:
+        super()._pre_attach(parent, name)
+        if self.name in parent.ds.variables:
+            raise KeyError(
+                f"parent {parent.name} already contains a variable named {self.name}"
+            )
+        path = _join_path(parent.path, name)
+        node_ds = self.to_dataset(inherited=False)
+        parent_ds = parent.to_dataset(inherited=False)
+        _check_alignment(path, node_ds, parent_ds, self.children)
+
+    def _post_attach_recursively(self: DataTree, parent: DataTree) -> None:
+        for k in parent._coord_names:
+            if k not in self._variables:
+                self._inherited_coord_names.add(k)
+
+        for child in self._children.values():
+            child._post_attach_recursively(self)
```

Comment:
After alignment has been checked, we build the list of inheritable coordinate names.